### ISSUE TITLE
[BUG / possible BUG] Handling of OCSP stapled responses using sha256 hash algorithm?

### ISSUE THREAD

**Issue by norrisjeremy** [drwetter/testssl.sh#2576]
The `testssl.sh` utility does not appear to work with stapled OCSP responses that use the `sha256` hash algorithm.

As seen [here](https://github.com/drwetter/testssl.sh/blob/541d3ff07a9e6ca81eaf7e52815ff229692277ed/testssl.sh#L2055), the constructed `openssl ocsp` command for verifying the stapled response does not specify a hash algorithm to use (i.e. no `-sha1`, `-sha256`, etc. param is specified).
Therefore the `openssl ocsp` command seems to assume a default value of `-sha1`.
This leads to an output that ends up like this if the OCSP stapled response does not actually use the `sha1` hash algorithm:
```
...
OCSP Request Data:
    Version: 1 (0x0)
    Requestor List:
        Certificate ID:
          Hash Algorithm: sha1
...
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
    Version: 1 (0x0)
...
    Responses:
    Certificate ID:
      Hash Algorithm: sha256
...
Response verify OK
/tmp/testssl.mKfCxE/host_certificate.pem: ERROR: No Status found.
```

This then leads to the `OCSP stapling` section to end up outputting the following:
```
 OCSP stapling                offered, error querying OCSP responder
```

If the constructed `openssl ocsp` command mentioned above is manually coerced to include a `-sha256` param, then the output verifies correct:
```
OCSP Request Data:
    Version: 1 (0x0)
    Requestor List:
        Certificate ID:
          Hash Algorithm: sha256
...
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
    Version: 1 (0x0)
...
    Responses:
    Certificate ID:
      Hash Algorithm: sha256
...
Response verify OK
/tmp/testssl.mKfCxE/host_certificate.pem: good
...
```

This in turn causes the `OCSP stapling` section to no longer indicate an error:
```
 OCSP stapling                offered, not revoked
```

Unfortunately, I am not sure how to best suggest a way to dynamically determine what hash algorithm the stapled OCSP response is using in order to correctly add the correct hash algorithm param.

**Comment by drwetter** [comment]
Hi @norrisjeremy ,

thanks! That should not too difficult to resolve....

As you didn't fill out the template I'd need two pieces of information: which openssl version were you using? Was this the 3.2 branch?

You might want to try ``--openssl=/usr/bin/openssl`` if you were using the supplied version of openssl-bad

**Comment by norrisjeremy** [comment]
Hi @drwetter,

I didn't include an OpenSSL version because this behavior isn't specific to any one version of OpenSSL: it occurs for all versions of OpenSSL.

Thanks,
Jeremy

**Comment by dcooper16** [comment]
Hi @norrisjeremy,

Thanks for reporting this. I just created a potential fix for this issue at https://github.com/dcooper16/testssl.sh/tree/sha256_stapled_ocsp. Basically, rather than assume the response uses SHA-1 in the Certificate ID it searches for SHA-1 (06052B0E03021A) in the response and then tries with SHA-1 if it is found. If SHA-1 isn't found or it doesn't work, then it searches for SHA-256 (0609608648016503040201) and tries with that if found.

The odds of finding 06052B0E03021A in the response if SHA-1 is not used for the Certificate ID is very small, so it should be very rare that the code actually tries to verify the response twice. For the same reason, we could easily add more hash functions to the list that it tries. However, at the moment I think it would be very unexpected for a stapled response to use something other than SHA-1 or SHA-256.

Note that I have not had a chance to test the code at https://github.com/dcooper16/testssl.sh/tree/sha256_stapled_ocsp at all. So, even if the basic idea behind the fix is correct, there is a good chance that I made a mistake in the implementation. I'll need to find time at some point in the future to actually test it.

**Comment by dcooper16** [comment]
This issue raises a similar, but somewhat more difficult to address, issue. In the case that `$PHONE_OUT` is true and an OCSP responder that is identified by URL in the authorityInfoAccess extension is being tested, what hash function should be used in the Certificate ID in the request? At the moment SHA-1 is used. But, do we need to support OCSP responders that can not process SHA-1 Certificate IDs? There is no way for testssl.sh to know what hash functions the OCSP responder accepts.

**Comment by drwetter** [comment]
@norrisjeremy : Can you name a host which would need the ``-sha256 param`` ? 

@dcooper16's patch looks good (thanks) but for him to test and for me to understand that information would be useful.

**Comment by norrisjeremy** [comment]
Hi @drwetter,

Unfortunately I cannot, as it was an internal-use (non-public) URL for our company and we were finally able to coerce the load balancer config for it to use SHA-1 instead of SHA-256.

Thanks,
Jeremy

**Comment by drwetter** [comment]
@ ANYONE else : do you know of a hosts with a stapled response using SHA256?

(Thanks @norrisjeremy )

**Comment by dcooper16** [comment]
I was able to test out my fix for this issue and so submitted a PR (#2589). Unfortunately I do not know of any publicly available server that provide stapled responses using SHA-256 either.

The OpenSSL s_server command has a `-status-file` option, which allowed me to tell it what OCSP response to staple. Thus, I could create an OCSP response with SHA-256 based CertID's and then use this option to staple it. This allowed me to test the PR.

As I noted, there is still the issue of what to do when `check_revocation_ocsp()` is following a URL and the OCSP responder doesn't support SHA-1 based CertIDs. This seems like something testssl.sh will need to address at some point, as https://datatracker.ietf.org/doc/draft-ietf-lamps-rfc5019bis/ encourages moving from SHA-1 to SHA-256.

At the moment, it seems that use of SHA-1 for CertID is still hard coded into implementations. This is what I found in Mozilla NSS for querying OCSP responders and also what I found in OpenSSL and Apache for obtaining a response to staple into a TLS handshake response message.

The only option I see is to try querying the OCSP responder using a SHA-1 based CertID and then trying again with SHA-256 if the first query failed. But, it's not a great solution since the query is relatively expensive. Fortunately, it is only done once and only if `$PHONE_OUT` is true, so it wouldn't really have a significant performance impact.
